### ISSUE TITLE
Type guards in array find

### ISSUE THREAD

**Issue by jer-sen** [microsoft/TypeScript#38945]
**TypeScript Version:**  3.9.2

**Search Terms:**

**Expected behavior:** No error

**Actual behavior:** Error `Type 'string | null | undefined' is not assignable to type 'string | undefined'.` Typescript doesn't take into account the type guard in find arrow function.

**Related Issues:**

**Code**
```ts
const res: string | undefined = ["a", null].find((v) => v !== null);
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
const res = ["a", null].find((v) => v !== null);

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?ssl=1&ssc=69&pln=1&pc=61#)

**Comment by MartinJohns** [comment]
`(v) => v !== null` is not a type guard. It's a function that returns `boolean`. A type could would return `v is string`.

You can fix it by adding a type annotation to make it a type guard: `find((v): v is string => v !== null)`

Alternative just add a comfort function:

    function isNotNull<T>(value: T | null): value is T { return value !== null; }
    
    find(isNotNull)

---

Also your playground link does not work.

**Comment by jer-sen** [comment]
`v !== null` is a type guard, no ?
Typescript should do the same assumption on `find` result than inside a `if (res !== null) { ... }`
Ideally a developer should not have to do anything special for this code to work. Don't you agree ?

But your code works fine thanks !

As for the playground link I used the "Export->Report GitHub issue on TypeScript" menu of typescript Playground...

**Comment by Barbiero** [comment]
> `v !== null` is a type guard, no ?
> Typescript should do the same assumption on `find` result than inside a `if (res !== null) { ... }`
> Ideally a developer should not have to do anything special for this code to work. Don't you agree ?
> 
> But your code works fine thanks !
> 
> As for the playground link I used the "Export->Report GitHub issue on TypeScript" menu of typescript Playground...

no, `v !== null` is an expression which typescript can use to infer a more restricte type, but it's not an actualy type guard.

The problem is that typescript can't really return guard-like expressions from a function without explicitly stating that the function is a guard. I've wrote the #37868 issue in an attempt to address that situation; hopefully someone more knowledgable than me can work that into the language.
